src/platform_linux/shmem.c

```diff
@@ -0,0 +1,1152 @@
+      hid);
+
+   if (!platform_valid_addr_in_heap(hid, ptr)) {
+      platform_error_log("[%s:%d::%s()] -> %s: Requesting to free memory"
+                         " at %p, for object '%s' which is a memory chunk not"
+                         " allocated from shared memory {start=%p, end=%p}.\n",
+                         file,
+                         lineno,
+                         func,
+                         __func__,
+                         ptr,
+                         objname,
+                         platform_shm_lop(hid),
+                         platform_shm_hip(hid));
+      return;
+   }
+
+   platform_shm_track_free(shminfo, ptr, objname, func, file, lineno);
+
+   if (Trace_shmem || Trace_shmem_frees) {
+      platform_default_log("  [%s:%d::%s()] -> %s: Request to free memory at "
+                           "%p for object '%s'.\n",
+                           file,
+                           lineno,
+                           func,
+                           __func__,
+                           ptr,
+                           objname);
+   }
+   return;
+}
+
+/*
+ * -----------------------------------------------------------------------------
+ * platform_shm_track_alloc() - Track the allocation of this large fragment.
+ * 'Tracking' here means we record this large-fragment in an array tracking
+ * large-memory fragments allocated.
+ * -----------------------------------------------------------------------------
+ */
+static void
+platform_shm_track_alloc(shmem_info *shm, void *addr, size_t size)
+{
+   debug_assert(
+      (size >= SHM_LARGE_FRAG_SIZE),
+      "Incorrect usage of this interface for requested size=%lu bytes."
+      " Size should be >= %lu bytes.\n",
+      size,
+      SHM_LARGE_FRAG_SIZE);
+
+   shm_lock_mem_frags(shm);
+
+   // Iterate through the list of memory fragments being tracked.
+   shm_frag_info *frag = shm->shm_mem_frags;
+   for (int fctr = 0; fctr < ARRAY_SIZE(shm->shm_mem_frags); fctr++, frag++) {
+      // If tracking entry points to a valid large-memory fragment ...
+      if (frag->shm_frag_addr) {
+         // As this is a newly allocated fragment being tracked, it should
+         // not be found elsewhere in the tracker array.
+         platform_assert((frag->shm_frag_addr != addr),
+                         "Error! Newly allocated large memory fragment at %p"
+                         " is already tracked at slot %d."
+                         " Fragment is allocated to PID=%d, to tid=%lu,"
+                         " and is %s (freed by PID=%d, by tid=%lu)\n.",
+                         addr,
+                         fctr,
+                         frag->shm_frag_allocated_to_pid,
+                         frag->shm_frag_allocated_to_tid,
+                         (frag->shm_frag_freed_by_pid ? "free" : "in use"),
+                         frag->shm_frag_freed_by_pid,
+                         frag->shm_frag_freed_by_tid);
+         continue;
+      }
+
```

Comment:
After the `continue`, the body of the for() will just process and do some book-keeping for a single fragment.
And then will `break`.

The unlock is immediately outside the `for()` loop.

I don't there is much to be gained by trying some micro-optimization and run the risk of race conditions creeping in because we unlocked too early.

No change done.